services/rpcfilters/api.go

```diff
@@ -0,0 +1,119 @@
+package rpcfilters
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/pborman/uuid"
+)
+
+type filter struct {
+	hashes []common.Hash
+	done   chan bool
+}
+
+func (f *filter) AddHash(hash common.Hash) {
+	if f.hashes == nil {
+		f.hashes = []common.Hash{hash}
+	} else {
+		f.hashes = append(f.hashes, hash)
+	}
+}
+
+func newFilter() *filter {
+	return &filter{
+		hashes: nil,
+		done:   make(chan bool),
+	}
+}
+
+// PublicAPI represents filter API that is exported to `eth` namespace
+type PublicAPI struct {
+	filters   map[rpc.ID]*filter
+	filtersMu sync.Mutex
+	event     *latestBlockChangedEvent
+}
+
+// NewPublicAPI returns a reference to the PublicAPI object
+func NewPublicAPI(event *latestBlockChangedEvent) *PublicAPI {
+	return &PublicAPI{
+		filters: make(map[rpc.ID]*filter),
+		event:   event,
+	}
+}
+
+// NewBlockFilter is an implemenation of `eth_newBlockFilter` API
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_newblockfilter
+func (api *PublicAPI) NewBlockFilter() rpc.ID {
+	api.filtersMu.Lock()
+	defer api.filtersMu.Unlock()
+
+	f := newFilter()
+	id := rpc.ID(uuid.New())
+
+	api.filters[id] = f
+
+	go func() {
+		id, s := api.event.Subscribe()
+		defer api.event.Unsubscribe(id)
+
+		for {
+			select {
+			case hash := <-s:
+				api.filtersMu.Lock()
+				f.AddHash(hash)
+				api.filtersMu.Unlock()
+			case <-f.done:
+				return
+			}
+		}
+
+	}()
+
+	return id
+}
+
+// UninstallFilter is an implemenation of `eth_uninstallFilter` API
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_uninstallfilter
+func (api *PublicAPI) UninstallFilter(id rpc.ID) bool {
+	api.filtersMu.Lock()
+	f, found := api.filters[id]
+	if found {
+		delete(api.filters, id)
+	}
+	api.filtersMu.Unlock()
+
+	if found {
+		close(f.done)
+	}
+
+	return found
+}
+
+// GetFilterChanges returns the logs for the filter with the given id since
+// last time it was called. This can be used for polling.
+//
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getfilterchanges
+func (api *PublicAPI) GetFilterChanges(id rpc.ID) (interface{}, error) {
```

Comment:
Why this return value is not `[]common.Hash`?

Reply:
Also, it says `returns the logs for the filter with the given id since last time it was called`. But hashes to `f.hashes` are only appended and never removed. This means it will return older hashes as well. Is it ok?

Reply:
they are removed. a few lines down

```
		hashes := f.hashes
		f.hashes = nil
```